PypeS/pype.py

```diff
@@ -167,127 +181,217 @@ def ParseArguments(self):
+
+        arguments:
+            memberEntry (obj): a memberEntry object
+            script (obj): a script object
+
+        returns:
+            (obj): a script member compatible with memberEntry, or None if not compatible
+        '''
         pushedInputMembers = [scriptMember for scriptMember in script.InputMembers if scriptMember.Pushed]
-        compatibleOutputMembers = [scriptMember for scriptMember in pushedInputMembers + script.OutputMembers if scriptMember.AutoPipe and (scriptMember.MemberName == member.MemberName) and (scriptMember.MemberType == member.MemberType)]
-        if not compatibleOutputMembers:
-            return None
-        return compatibleOutputMembers[0]
-    
+        for scriptMember in pushedInputMembers + script.OutputMembers:
+            if not scriptMember.AutoPipe: continue # only true when --no-auto flag is provided
+            if ((scriptMember.MemberName == memberEntry.MemberName) and 
+                (scriptMember.MemberType == memberEntry.MemberType)):
+                # only if scriptMember name/types both match memberEntry names/types
+                return scriptMember
+        return None
+
     def AutoPipeScriptObject(self,scriptObject):
+        '''automatically connect inputs members of a scriptObject with outputMembers which have come before.
+
+        When a script completes execution, the next script to execute is sent here. This attempts to automatically
+        connect compatible output members of prior script objects with the input members of this (the next) script
+        object. If a compatible member is found, this script object's input memberEntry.MemberPipe attribute
+        is set. Note: Automatic piping is override by explicit member piping.
+
+        example:
+            for -> vmtkimagereader -ifile foo --pipe vmtkimageviewer
+                the vmtkimageviewer memberEntry.MemberPipe for memberEntry MemberName == Image
+                is set to: vmtkimagereader-0.Image
+        '''
         self.PrintLog('Automatic piping ' + scriptObject.ScriptName)
         for memberEntry in scriptObject.InputMembers:
-            if not memberEntry.AutoPipe:
-                continue
-            if memberEntry.MemberType == 'id':
-                continue
-            if memberEntry.MemberType == 'handle':
+            if ((not memberEntry.AutoPipe) or
+                (memberEntry.MemberType == 'id') or
+                (memberEntry.MemberType == 'handle')): continue # these member types we never want to match
+            # if a compatible member is found between InputMember entries and the prior script objects output
+            # members, the compatible member is assigned to the pipedScriptObject name. We use the None value
+            # to check wheather a compatible memver was found or not.
+            pipedScriptObject = None
+            for candidateScript in self.ScriptObjectList:
+                if self._CompatibleMember(memberEntry,candidateScript):
```

Comment:
set candidate pipedMember outside the if condition and check if not None, to avoid calling _CompatibleMember twice